File: android/sdk/wear/core/src/main/java/app/organicmaps/sdk/wear/WearBridge.java
package app.organicmaps.sdk.wear;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import app.organicmaps.wear.protocol.WearNavigationState;

/**
 * Routes navigation state to a paired Wear OS device.
 *
 * <p>The Google variant bundles {@code :sdk:wear:gms}, whose manifest-merged {@code ContentProvider}
 * registers the real publisher at startup. On F-Droid/Huawei/Web nothing registers and the bridge
 * falls back to a no-op, so callers in common code never reference Play services types.
 */
public final class WearBridge
{
  @NonNull
  private static final WearNavigationPublisher NO_OP = state -> {};

  @Nullable
  private static volatile WearNavigationPublisher sPublisher;

  /** Called once at process startup from the Google Wear module. */
  public static void register(@NonNull WearNavigationPublisher publisher)
  {
    sPublisher = publisher;
  }

  /** Never null: the registered publisher, or a no-op fallback when no Wear bridge is present. */
  @NonNull
  private static WearNavigationPublisher publisher()
  {
    final WearNavigationPublisher publisher = sPublisher;
    return publisher != null ? publisher : NO_OP;
  }

  /** Maps a navigation-active flag to a published state; suitable as a routing state listener. */
  public static void publishNavigating(boolean navigating)
  {
    publisher().publish(navigating ? WearNavigationState.navigation() : WearNavigationState.normal());
  }

  private WearBridge() {}
}

File: android/sdk/wear/core/src/main/java/app/organicmaps/sdk/wear/WearBridge.kt
package app.organicmaps.sdk.wear

import app.organicmaps.wear.protocol.WearNavigationDetails
import app.organicmaps.wear.protocol.WearNavigationMode

/**
 * Routes navigation data to a paired Wear OS device.
 *
 * Google debug and beta bundle `:sdk:wear:gms`, whose manifest-merged `ContentProvider`
 * registers the real publishers at startup. Everywhere else -- Google release and profileable,
 * F-Droid, Huawei, Web -- nothing registers and the bridge falls back to no-ops, so callers in
 * common code never reference Play services types.
 */
object WearBridge {
    private val NO_OP_MODE = WearNavigationPublisher {}
    private val NO_OP_DETAILS = WearNavigationDetailsPublisher {}

    @Volatile
    private var modePublisher: WearNavigationPublisher = NO_OP_MODE

    @Volatile
    private var detailsPublisher: WearNavigationDetailsPublisher = NO_OP_DETAILS

    /** Called once at process startup from the Google Wear module. */
    fun register(publisher: WearNavigationPublisher) {
        modePublisher = publisher
    }

    /** Called once at process startup from the Google Wear module. */
    fun registerDetails(publisher: WearNavigationDetailsPublisher) {
        detailsPublisher = publisher
    }

    /** Maps a navigation-active flag to a published persistent mode. */
    @JvmStatic
    fun publishNavigating(navigating: Boolean) {
        val mode =
            if (navigating) WearNavigationMode.NAVIGATION else WearNavigationMode.NORMAL

        modePublisher.publish(mode)
    }

    /** Publishes transient details while navigation is active. */
    @JvmStatic
    fun publishDetails(details: WearNavigationDetails) {
        detailsPublisher.publish(details)
    }
}
